### PR TITLE
feat: queue optimizing backend implementation

### PR DESCRIPTION
**Pull request by miguel-crespo-fdc**
Backend Implementation for queue optimizing. Adds a variable and configurable length to the queue and and drops requests once the queue limit is reached.

### CODE REVIEW

File: services/cd-service/pkg/repository/repository.go [hunk 1]
```diff
@@ -317,6 +319,16 @@ func GetTags(cfg RepositoryConfig, repoName string, ctx context.Context) (tags [
 	return tags, nil
 }
 
+// Does not spawn the background action, namely the queue processing, for testing
+func NewAuxiliary(ctx context.Context, cfg RepositoryConfig) (Repository, error) {
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/repository.go:323]: can we move this to repository_test.go?

File: services/cd-service/pkg/service/batch.go [hunk 2]
```diff
@@ -332,6 +332,12 @@ func (d *BatchServer) ProcessBatch(
 				}
 				return &api.BatchResponse{Results: errorResults}, nil
 			}
+		case *repository.CreateBatchError:
+			if transformerError.GetType() == repository.QUEUE_IS_FULL {
+				return nil, status.Error(codes.ResourceExhausted, "Could not process ProcessBatch request. Too many pending operations. Queue is full.")
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/service/batch.go:337]: I'd like to add the configured maximum to the error message here.

File: services/cd-service/pkg/repository/repository_test.go [hunk 3]
```diff
@@ -2078,3 +2079,129 @@ func TestSendWebhookToArgoCd(t *testing.T) {
...
+						return func() error {
+							return nil
+						}
+					}
+					repo.(*repository).ProcessQueueOnce(ctx, e, dummyPushUpdateFunction, dummyPushActionFunction)
+				default:
+				}
+				select {
+				case err := <-errCh:
+					t.Log(err)
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/repository_test.go:2150]: this looks like will log `nil` here.

**Review by sven-urbanski-freiheit-com — CHANGES_REQUESTED** [review]
requesting some small changes.
I would also like @hannesg to take a look

File: services/cd-service/pkg/repository/queue.go [hunk 4]
```diff
@@ -54,12 +55,23 @@ func (q *queue) add(ctx context.Context, transformers []Transformer) <-chan erro
...
+
 	select {
 	case q.transformerBatches <- e:
 		return resultChannel
 	case <-ctx.Done():
 		e.finish(ctx.Err())
 		return resultChannel
+	default:
+		//Channel is full, we don't want to put anything else there.
+		e.finish(&TransformerBatchApplyError{
```
Comment (hannesg) [inline services/cd-service/pkg/repository/queue.go:67]: I think you can have this much simpler here.
Create an error on top-level like this:

```go
var ErrQueueFull error = fmt.Errorf("queue full")
```

Then you can use `e.finish(ErrQueueFull)` here and later when checking [here](https://github.com/freiheit-com/kuberpult/pull/1466/files#diff-1a2c7d4b2d78782768dc4449dbb8e6fb964eaf7a97d631faebd14adaec1fb6d9L322) you can use this:

```
if errors.Is(err, ErrQueueFull) {
   return nil, status.Error(codes.ResourceExhausted, "Could not process ProcessBatch request. Too many pending operations. Queue is full.")
}
```

File: services/cd-service/pkg/cmd/server.go [hunk 5]
```diff
@@ -181,6 +182,11 @@ func RunServer() {
 				zap.String("details", "the maximum number of commits per push must be at least 1"),
 			)
 		}
+		if c.MaximumQueueSize < 1 || c.MaximumQueueSize > 100 {
+			logger.FromContext(ctx).Fatal("cd.config",
+				zap.String("details", "the size of the queue must be between 2 and 100"),
```
Comment (jonaswittbrodt) [inline services/cd-service/pkg/cmd/server.go:187]: This error message does not match the if statement.

**Review by hannesg — APPROVED** [review]

**Review by sven-urbanski-freiheit-com — APPROVED** [review]

**Pull request merged**